From 491eadf82ac461a44df71c0c27b3619d4084abfc Mon Sep 17 00:00:00 2001
From: Jackal Guo <jackalguo@google.com>
Date: Wed, 11 Jan 2023 14:36:32 +0800
Subject: [PATCH] [RESTRICT AUTOMERGE] Limit length of the name in
 <uses-permission>

To mitigate a boot loop with a huge name string, this CL restricts
the max length of the name in <users-permission> attr.

Bug: 259942609
Test: manually using the PoC in the buganizer to ensure the symptom
      no longer exists.
Change-Id: I35dde9d6bea9eb7c1b81e8cbadeeb908b3fb713f
(cherry picked from commit de8ef32d020ce4efe5dcaae09c9b8e0cf7efb2db)
Merged-In: I35dde9d6bea9eb7c1b81e8cbadeeb908b3fb713f
---
 .../server/pm/pkg/parsing/ParsingPackageUtils.java        | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/services/core/java/com/android/server/pm/pkg/parsing/ParsingPackageUtils.java b/services/core/java/com/android/server/pm/pkg/parsing/ParsingPackageUtils.java
index 06a54a461d5e..7289fdf245a2 100644
--- a/services/core/java/com/android/server/pm/pkg/parsing/ParsingPackageUtils.java
+++ b/services/core/java/com/android/server/pm/pkg/parsing/ParsingPackageUtils.java
@@ -245,6 +245,9 @@ public class ParsingPackageUtils {
     private static final String MAX_NUM_COMPONENTS_ERR_MSG =
             "Total number of components has exceeded the maximum number: " + MAX_NUM_COMPONENTS;
 
+    /** The maximum permission name length. */
+    private static final int MAX_PERMISSION_NAME_LENGTH = 512;
+
     @IntDef(flag = true, prefix = { "PARSE_" }, value = {
             PARSE_CHATTY,
             PARSE_COLLECT_CERTIFICATES,
@@ -1271,6 +1274,11 @@ public class ParsingPackageUtils {
             // that may change.
             String name = sa.getNonResourceString(
                     R.styleable.AndroidManifestUsesPermission_name);
+            if (TextUtils.length(name) > MAX_PERMISSION_NAME_LENGTH) {
+                return input.error(INSTALL_PARSE_FAILED_MANIFEST_MALFORMED,
+                        "The name in the <uses-permission> is greater than "
+                                + MAX_PERMISSION_NAME_LENGTH);
+            }
 
             int maxSdkVersion = 0;
             TypedValue val = sa.peekValue(
-- 
2.39.1.456.gfc5497dd1b-goog

